web/pandas/pdeps/0009-io-extensions.md

```diff
@@ -0,0 +1,395 @@
+pandas internals happen. It would also allow other possible consumers of the
+connectors to not have pandas as a dependency. Testing also becomes simpler by
+using Apache Arrow, since connectors can be tested independently, and pandas does
+not need to be tested for each connector. If the Apache Arrow specification is
+respected in both sides, the communication between connectors and pandas is
+guaranteed to work. If pandas eventually has a hard dependency on an Apache
+Arrow implementation, this should be the preferred interface.
+
+Allowing connectors to use pandas dataframes directly makes users not have to
+depend on PyArrow for connectors that do not use an Apache Arrow object. It also
+helps move existing connectors to this new API, since they are using pandas
+dataframes as an exchange object now. It has the disadvantages stated in the
+previous paragraph, and a future proposal may be created to discuss deprecating
+pandas dataframes as a possible interface for connectors.
+
+In case a `read` method returned something different than a PyArrow table,
+pandas would raise an exception. pandas would expect all `to_` methods to have
+`table: pyarrow.Table | pandas.DataFrame` as the first parameter, and it would
+raise an exception otherwise. The `table` parameter would be exposed as the
+`self` parameter of the `to_*` method in pandas.
+
+In case the Apache Arrow interface is used, metadata not supported by Apache
+Arrow may be provided by users. For example the column to use for row indices
+or the data type backend to use in the object being created. This would be
+managed independently from the connectors. Given the previous example, a new
+argument `index_col` could be added directly into pandas to the function or
+method automatically generated from the entrypoint. Since this would apply to
+all functions and methods automatically generated, it would also improve the
+consistency of pandas connectors. For example:
+
+```python
+def read_myformat(index_col=None, *args, **kwargs):
+    # The third-party connector doesn't need to know about functionality
+    # specific to pandas like the row index
+    arrow_table = third_party_connector.from_myformat(*args, **kwargs)
+
+    df = convert_arrow_table_to_pandas_dataframe(arrow_table)
+
+    # Transformations to the dataframe with custom parameters is possible
+    if index_col is not None:
+        df = df.set_index(index_col)
+
+    return df
+```
+
+#### Connector guidelines
+
+In order to provide a better and more consistent experience to users, guidelines
+will be created to unify terminology and behavior. Some of the topics to unify are
+defined next.
+
+**Existence and naming of columns**, since many connectors are likely to provide
+similar features, like loading only a subset of columns in the data, or dealing
+with file names. Examples of recommendations to connector developers:
+
+- `columns`: Use this argument to let the user load a subset of columns. Allow a
```

Comment:
we should have specific requirements here rather than just guidelines

otherwise this is going to be a mess